src/scribe_data/check/check_query_forms.py

```diff
@@ -183,14 +198,251 @@ def return_correct_form_label(qids: list):
+        return_forms = []
+        for part in match[1].split():
+            if "?" in part:
+                form = part.split("?")[-1]
+                if "AS" in form:
+                    form = form.split("AS")[0].strip()
+                return_forms.append(form)
+
+        unique_forms = set(return_forms)
+        if len(return_forms) != len(unique_forms):
+            error_output += f"\nDuplicate forms found: {', '.join([form for form in return_forms if return_forms.count(form) > 1])}"
+            return error_output
+
+        return True
+
+    return True
+
+
+# MARK: Unreturned Forms
+
+
+def check_unreturned_optional_forms(query_text: str) -> str:
+    """
+    Checks if there are any optional forms in the query that aren't returned in the SELECT statement.
+
+    Parameters
+    ----------
+    query_text : str
+        The full text of the SPARQL query.
+
+    Returns
+    -------
+    str
+        Error message listing any unreturned forms, or empty string if all forms are returned.
+    """
+    # Extract forms from SELECT statement.
+    select_pattern = r"SELECT\s*(.*?)\s*WHERE"
+    select_forms = set()
+    if select_match := re.search(
+        pattern=select_pattern, string=query_text, flags=re.DOTALL
+    ):
+        for part in select_match[1].split():
+            if "?" in part:
+                form = part.split("?")[-1]
+                if "AS" in form:
+                    form = form.split("AS")[0].strip()
+                select_forms.add(form)
+
+    # Extract forms from OPTIONAL blocks
+    optional_forms = set()
+    optional_pattern = r"OPTIONAL\s*\{([^}]*)\}"
+    for match in re.finditer(optional_pattern, query_text):
+        form_text = match.group(1)
+        rep_pattern = r"ontolex:representation\s+\?([\w]+)\s*;"
+        if rep_match := re.search(rep_pattern, form_text):
+            optional_forms.add(rep_match[1])
+
+    # Find forms that appear in OPTIONAL blocks but not in SELECT.
+    unreturned_forms = optional_forms - select_forms
+
+    if unreturned_forms:
+        return f"Unreturned optional forms: {', '.join(sorted(unreturned_forms))}"
+
+    return ""
+
+
+# MARK: Undefined Forms
+
+
+def check_undefined_return_forms(query_text: str) -> str:
+    """
+    Checks if the query is trying to return forms that aren't defined in the WHERE clause
+    when there are no OPTIONAL blocks.
+
+    Parameters
+    ----------
+        query_text : str
+            The full text of the SPARQL query.
+
+    Returns
+    -------
+        str
+            Error message listing any undefined forms being returned, or empty string if all
+            returned forms are properly defined.
+    """
+
+    # Check if query has any OPTIONAL blocks.
+    optional_pattern = r"OPTIONAL\s*\{"
+    has_optional_blocks = bool(re.search(optional_pattern, query_text))
+
+    if has_optional_blocks:
+        return ""  # skip check for queries with OPTIONAL blocks
+
```

Comment:
I’m sorry, but in this part I don't understand why you guyes @andrewtavis @axif0 skipped queries that contain the `OPTIONAL` statement:

```python
# Check if query has any OPTIONAL blocks.
optional_pattern = r"OPTIONAL\s*\{"
has_optional_blocks = bool(re.search(optional_pattern, query_text))

if has_optional_blocks:
    return ""  # skip check for queries with OPTIONAL blocks
```

Almost all the queries that have been or will be written will include an `OPTIONAL` statement, and this is where many people (including myself) often do errors.